samples/join-a-team-using-qr-code/csharp/JoinTeamByQR/TokenExchangeHelper.cs

```diff
@@ -0,0 +1,162 @@
+        {
+            if (turnContext.Activity.Name != SignInConstants.TokenExchangeOperationName)
+            {
+                throw new InvalidOperationException("Only 'signin/tokenExchange' invoke activities can be procssed by TokenExchangeHelper.");
+            }
+
+            if (!await this.ExchangedTokenAsync(turnContext, cancellationToken).ConfigureAwait(false))
+            {
+                // If the TokenExchange is NOT successful, the response will have already been sent by ExchangedTokenAsync
+                return false;
+            }
+
+            // If a user is signed into multiple Teams clients, the Bot might receive a "signin/tokenExchange" from each client.
+            // Each token exchange request for a specific user login will have an identical Activity.Value.Id.
+            // Only one of these token exchange requests should be processe by the bot.  For a distributed bot in production,
+            // this requires a distributed storage to ensure only one token exchange is processed.
+
+            // This example utilizes Bot Framework IStorage's ETag implementation for token exchange activity deduplication.
+
+            // Create a StoreItem with Etag of the unique 'signin/tokenExchange' request
+            var storeItem = new TokenStoreItem
+            {
+                ETag = (turnContext.Activity.Value as JObject).Value<string>("id")
+            };
+
+            var storeItems = new Dictionary<string, object> { { TokenStoreItem.GetStorageKey(turnContext), storeItem } };
+            try
+            {
+                // Writing the IStoreItem with ETag of unique id will succeed only once
+                await _storage.WriteAsync(storeItems, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+                // Memory storage throws a generic exception with a Message of 'Etag conflict. [other error info]'
+                when (ex.Message.StartsWith("Etag conflict"))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private async Task<bool> ExchangedTokenAsync(ITurnContext turnContext, CancellationToken cancellationToken)
+        {
+            TokenResponse tokenExchangeResponse = null;
+            var tokenExchangeRequest = ((JObject)turnContext.Activity.Value)?.ToObject<TokenExchangeInvokeRequest>();
+
+            try
+            {
+                tokenExchangeResponse = await (turnContext.Adapter as IExtendedUserTokenProvider).ExchangeTokenAsync(
+                    turnContext,
+                    _oAuthConnectionName,
+                    turnContext.Activity.From.Id,
+                    new TokenExchangeRequest { Token = tokenExchangeRequest.Token },
+                    cancellationToken).ConfigureAwait(false);
+            }
+#pragma warning disable CA1031 // Do not catch general exception types (ignoring, see comment below)
```

Comment:
Why do we need these?

Reply:
Those are included to ignore exceptions. The format has been used from old samples so implemented the same here